### PR TITLE
Add CLI interface for parsing and validation

### PR DESCRIPTION
**Pull request by gelraen**

### CODE REVIEW

**Comment by vstakhov** [comment]
That looks good. Have you checked this project as well by the way: https://github.com/allanjude/uclcmd

**Comment by gelraen** [comment]
`uclcmd` looks a bit more complicated than just a CLI wrapper for libucl. And I don't see there an option for performing validation.

**Comment by vstakhov** [comment]
I don't see either :) I've just reminded this project as somehow related.

File: utils/Makefile.am [hunk 1]
```diff
@@ -11,8 +11,12 @@ ucl_objdump_SOURCES = objdump.c
 ucl_objdump_LDADD = $(common_utils_ldadd)
 ucl_objdump_CFLAGS = $(common_utils_cflags)
 
+ucl_SOURCES = ucl.c
+ucl_LDADD = $(common_utils_ldadd)
+ucl_CFLAGS = $(common_utils_cflags)
+
 if UTILS
-UTL = ucl_chargen ucl_objdump
+UTL = ucl_chargen ucl_objdump ucl
```
Comment (vstakhov) [inline utils/Makefile.am]: Could you please use less ambiguous name? E.g. ucl-tool or ucl-valdator (same for source file).

File: utils/ucl.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <getopt.h>
+#include <sysexits.h>
```
Comment (vstakhov) [inline utils/ucl.c]: That is BSD specific include.

File: utils/ucl.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      exit(EX_USAGE);
+      break;
+    }
+  }
+
+  parser = ucl_parser_new(0);
+  buf = malloc(BUFSIZ);
+  size = BUFSIZ;
+  while(!feof(in) && !ferror(in)) {
+    if (r == size) {
```
Comment (vstakhov) [inline utils/ucl.c]: As far as I see, `r` is not initialized here.

File: utils/ucl.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+int main(int argc, char **argv) {
+  char ch;
+  FILE *in = stdin, *out = stdout;
+  const char *schema = NULL;
+  unsigned char *buf = NULL;
+  size_t size = 0;
+  struct ucl_parser *parser = NULL;
+  int r;
```
Comment (vstakhov) [inline utils/ucl.c]: `r` is `int` whilst `size` is `size_t`. That seems to be dangerous and irrational, since `fread` returns `size_t`.

**Comment by bapt** [comment]
FYI uclcmd plans to add validation I pushed that into @allanjude TODO list back in AsiaBSDCon

File: utils/ucl.c (hunk 4 above)
Comment (gelraen) [inline utils/ucl.c]: Agreed. Switched to size_t.

File: utils/ucl.c (hunk 3 above)
Comment (gelraen) [inline utils/ucl.c]: Fixed.

File: utils/ucl.c (hunk 2 above)
Comment (gelraen) [inline utils/ucl.c]: Fixed.

**Pull request merged**